lib/web/recordingplayback.go

```diff
@@ -0,0 +1,653 @@
+	clt              events.SessionStreamer
+	sessionID        string
+	logger           *slog.Logger
+	mu               sync.Mutex
+	cancelActiveTask context.CancelFunc
+	wg               sync.WaitGroup
+	ws               *websocket.Conn
+	writeChan        chan websocketMessage
+	closeSent        bool // tracks if a close message has been sent
+	stream           recordingStream
+	terminal         recordingTerminal
+}
+
+// fetchRequest represents a request for session events.
+type fetchRequest struct {
+	requestType          requestType
+	startOffset          time.Duration
+	endOffset            time.Duration
+	requestID            int
+	requestCurrentScreen bool
+}
+
+// sessionEvent represents a single session event with its type, timestamp, and data.
+type sessionEvent struct {
+	eventType  responseType
+	timeOffset time.Duration
+	data       []byte
+}
+
+func (h *Handler) recordingPlaybackWS(
+	w http.ResponseWriter,
+	r *http.Request,
+	p httprouter.Params,
+	sctx *SessionContext,
+	cluster reversetunnelclient.Cluster,
+	ws *websocket.Conn,
+) (interface{}, error) {
+	sessionID := p.ByName("session_id")
+	if sessionID == "" {
+		h.closeWebsocketWithError(r.Context(), ws, sessionID, trace.BadParameter("missing session ID in request URL"))
+		return nil, nil
+	}
+
+	ctx := r.Context()
+	clt, err := sctx.GetUserClient(ctx, cluster)
+	if err != nil {
+		h.closeWebsocketWithError(ctx, ws, sessionID, trace.Wrap(err, "failed to get user client"))
+		return nil, nil
+	}
+
+	playback := newRecordingPlayback(ctx, ws, clt, sessionID, h.logger)
+
+	playback.run()
+
+	return nil, nil
+}
+
+func (h *Handler) closeWebsocketWithError(ctx context.Context, ws *websocket.Conn, sessionID string, err error) {
+	data := []byte(err.Error())
+
+	totalSize := responseHeaderSize + len(data)
+	buf := make([]byte, totalSize)
+
+	encodeEvent(buf, 0, eventTypeError, 0, data, 0)
+
+	if err := ws.WriteMessage(websocket.BinaryMessage, buf); err != nil {
+		h.logger.ErrorContext(ctx, "failed to send event", "session_id", sessionID, "error", err)
+	}
+
+	deadline := time.Now().Add(websocketCloseTimeout)
+
+	// Send close frame to initiate graceful shutdown
+	if err := ws.SetWriteDeadline(deadline); err != nil {
+		h.logger.DebugContext(ctx, "failed to set write deadline", "session_id", sessionID, "error", err)
+	}
+	if err := ws.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")); err != nil {
+		h.logger.DebugContext(ctx, "failed to send close message", "session_id", sessionID, "error", err)
+	}
+
+	// Wait for peer's close frame response (or timeout)
+	if err := ws.SetReadDeadline(deadline); err != nil {
+		h.logger.DebugContext(ctx, "failed to set read deadline", "session_id", sessionID, "error", err)
+	}
+
+	// Log if we got something other than a close acknowledgement
+	if _, _, err := ws.ReadMessage(); err != nil && !websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
```

Comment:
Won't this conflict with `readLoop` which is also concurrently calling ReadMessage?